common/src/main/java/tachyon/replay/ReplayCache.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Licensed to the University of California, Berkeley under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.replay;
+
+import java.io.IOException;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.base.Throwables;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+
+import tachyon.exception.TachyonException;
+import tachyon.thrift.TachyonTException;
+import tachyon.thrift.ThriftIOException;
+
+/**
+ * An RPC cache which uses RPC ids to avoid repeating non-idempotent RPCs due to retries.
+ *
+ * Whenever a {@link RetryCallable} is run via this class, the RPC id and return value for the
+ * {@link RetryCallable} are remembered so that if the RPC id is replayed while the id is still in
+ * the cache, the return value can be immediately returned without executing the
+ * {@link RetryCallable}.
+ *
+ * For RPCs which may throw {@link IOException}, use {@link ReplayCallableThrowsIOException}.
+ *
+ * Example usage:<br>
+ *
+ * <pre>
+ * <code>
+ * private final Cache<String, Long> cache = ReplayCache.create();
```

Comment:
These are convenient since they let you write 
```
private final Cache<String, Long> cache = ReplayCache.newReplayCache();
```
instead of
```
private final Cache<String, Long> cache = new ReplayCache<String, Long>();
```

They also give us more flexibility - we could in the future return different types for different arguments, we don't need to use ```whenNew()`` to mock them, and Joshua Bloch likes them http://www.informit.com/articles/article.aspx?p=1216151